Sources/Controllers/AudioController.swift

```diff
@@ -0,0 +1,172 @@
+            }
+        }
+    }
+
+    internal func pauseSound(for message: MessageType, in audioCell: AudioMessageCell) {
+        audioPlayer?.pause()
+        audioCell.playButton.isSelected = false // show play button on audio cell since the sound is on pause
+        progressTimer?.invalidate()
+        // call stop audio delegate
+        if let cell = playingCell {
+            cell.delegate?.didPauseAudio(in: cell)
+        }
+    }
+
+    internal func stopAudioInCurrentPlayingCell() {
+        // call stop audio delegate
+        if let cell = playingCell {
+            cell.delegate?.didStopAudio(in: cell)
+        }
+        // stop audio
+        audioPlayer?.stop()  // stop method did not call audioPlayerDidFinishPlaying(player:,flag:)
+        audioPlayerDidFinishPlaying(audioPlayer ?? AVAudioPlayer(), successfully: true)
+    }
+
+    // MARK: - Fire Methods
+    @objc private func didFireProgressTimer(_ timer: Timer) {
+        guard let player = audioPlayer, let collectionView = messageCollectionView, let cell = playingCell else {
+            return
+        }
+        // check if can update playing cell
+        if let playingCellIndexPath = collectionView.indexPath(for: cell) {
+            // 1. get the current message that decorates the playing cell
+            // 2. check if current message is the same with playing message, if so then update the cell content
+            // Note: Those messages differ in the case of cell reuse
+            let currentMessage = collectionView.messagesDataSource?.messageForItem(at: playingCellIndexPath, in: collectionView)
+            if currentMessage != nil && currentMessage?.messageId == playingMessage?.messageId {
+                // messages are the same updte cell content
+                cell.progressView.progress = (player.duration == 0) ? 0 : Float(player.currentTime/player.duration)
+                cell.durationLabel.text = AudioMessageCell.durationString(from: Float(player.currentTime))
+            }
+        }
+    }
+
+    // MARK: - Private Methods
+    private func startProgressTimer() {
+        // reset timer first
+        progressTimer?.invalidate()
+        progressTimer = nil
+        // start timer
+        progressTimer = Timer.scheduledTimer(timeInterval: 0.1, target: self, selector: #selector(AudioController.didFireProgressTimer(_:)), userInfo: nil, repeats: true)
+    }
+
+    private func resumeSound() {
+        audioPlayer?.play()
+        playingCell?.playButton.isSelected = true // show pause button on audio cell since the sound is currently playing
+        startProgressTimer()
+        // call start audio delegate
+        if let cell = playingCell {
+            cell.delegate?.didStartAudio(in: cell)
+        }
+    }
+
+}
+
+// MARK: - AVAudioPlayerDelegate
+extension AudioController: AVAudioPlayerDelegate {
+
+    internal func audioPlayerDidFinishPlaying(_ player: AVAudioPlayer, successfully flag: Bool) {
```

Comment:
The audioPlayerDidFinishPlaying(::) is a delegate method that notify when AVAudioPlayer finished play. I can not modify the structure of this method since this one is from the OS system. I think is better to implement a new method without parameters an call it from audioPlayerDidFinishPlaying(::) and stopAudioInCurrentPlayingCell() methods.